mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+        self.sqrt_recip_alplas_bar = np.sqrt(1.0 / self.alphas_bar)
+        self.sqrt_recipm1_alphas_bar = np.sqrt(1.0 / self.alphas_bar - 1)
+
+        # calculations for posterior q(x_{t-1} | x_t, x_0)
+        self.tilde_betas_t = self.betas * (1 - self.alphas_bar_prev) / (
+            1 - self.alphas_bar)
+        # clip log var for tilde_betas_0 = 0
+        self.log_tilde_betas_t_clipped = np.log(
+            np.append(self.tilde_betas_t[1], self.tilde_betas_t[1:]))
+        self.tilde_mu_t_coef1 = np.sqrt(
+            self.alphas_bar_prev) / (1 - self.alphas_bar) * self.betas
+        self.tilde_mu_t_coef2 = np.sqrt(
+            self.alphas) * (1 - self.alphas_bar_prev) / (1 - self.alphas_bar)
+
+    def get_betas(self):
+        """Get betas by defined schedule method in diffusion process."""
+        self.betas_schedule = self.betas_cfg.pop('type')
+        if self.betas_schedule == 'linear':
+            return self.linear_beta_schedule(self.num_timesteps,
+                                             **self.betas_cfg)
+        elif self.betas_schedule == 'cosine':
+            return self.cosine_beta_schedule(self.num_timesteps,
+                                             **self.betas_cfg)
+        else:
+            raise AttributeError(f'Unknown method name {self.beta_schedule}'
+                                 'for beta schedule.')
+
+    @staticmethod
+    def linear_beta_schedule(diffusion_timesteps, beta_0=1e-4, beta_T=2e-2):
+        r"""Linear schedule from Ho et al, extended to work for any number of
+        diffusion steps.
+
+        Args:
+            diffusion_timesteps (int): The number of betas to produce.
+            beta_0 (float, optional): `\beta` at timestep 0. Defaults to 1e-4.
+            beta_T (float, optional): `\beta` at timestep `T` (the final
+                diffusion timestep). Defaults to 2e-2.
+
+        Returns:
+            np.ndarray: Betas used in diffusion process.
+
+        """
+        scale = 1000 / diffusion_timesteps
+        beta_0 = scale * beta_0
+        beta_T = scale * beta_T
+        return np.linspace(
+            beta_0, beta_T, diffusion_timesteps, dtype=np.float64)
+
+    @staticmethod
+    def cosine_beta_schedule(diffusion_timesteps, max_beta=0.999, s=0.008):
+        r"""Create a beta schedule that discretizes the given alpha_t_bar
+        function, which defines the cumulative product of `(1-\beta)` over time
+        from `t = [0, 1]`.
+
+        Args:
+            diffusion_timesteps (int): The number of betas to produce.
+            max_beta (float, optional): The maximum beta to use; use values
+                lower than 1 to prevent singularities. Defaults to 0.999.
+            s (float, optional): Small offset to prevent `\beta` from being too
+                small near `t = 0` Defaults to 0.008.
+
+        Returns:
+            np.ndarray: Betas used in diffusion process.
+
+        """
+
+        def f(t, T, s):
```

Comment:
This name is not an official name and please give docs for this function.

Reply:
You may further check the implementation of this function. Do you want a `staticemthod`?

Reply:
How about move `get_betas`, `linear_beta_schedule` and `cosine_beta_schedule` (this function) to `mmgen/models/diffusions/utils.py` ?